### PR TITLE
Refactor to use aiobotocore, python3 and tornado 7

### PR DESCRIPTION
**Pull request by kkopachev**

### CODE REVIEW

File: tc_aws/storages/s3_storage.py [hunk 1]
```diff
@@ -23,53 +27,146 @@ def __init__(self, context):
...
+
+        file_key = await file_key['Body'].read()
+
+        return file_key.decode('utf-8')
+
+    async def get_detector_data(self, path):
+        """
+        Retrieves detector data from storage
+        :param string path: Path where the data is stored
+        :param callable callback: Callback function for once the retrieval is done
```
Comment (heynemann) [inline tc_aws/storages/s3_storage.py]: All comments with callback can go, right?

File: tc_aws/storages/s3_storage.py [hunk 2]
```diff
@@ -23,53 +27,146 @@ def __init__(self, context):
...
-            if key is None or self._get_error(key) is not None:
-                callback(None)
-            else:
-                callback(key['Body'].read())
+        try:
+            file = await super(Storage, self).get(path)
+        except BotoCoreError:
+            return None
+
+        return await file['Body'].read()
```
Comment (heynemann) [inline tc_aws/storages/s3_storage.py]: In the above method you check to see whether Body exists. Wouldn't it be better to have a `read_object` function that does this logic and just call it everywhere?

File: tests/test_result_storage.py [hunk 3]
```diff
@@ -26,83 +24,48 @@ class Request(object):
 class S3StorageTestCase(S3MockedAsyncTestCase):
 
     @gen_test
-    def test_can_store_image(self):
```
Comment (heynemann) [inline tests/test_result_storage.py]: Why did you delete both tests? They do not apply anymore?

**Comment by Bladrak** [comment]
Thanks for this huge update! :) 
This should probably go to a new branch on this repo, while Thumbor's 7th version is in alpha.
I'm also wondering how we can test for compatibility with previous versions (noticeably for potential issues on image paths).
This will probably also require some performance tests as have been done in the current version, as you're removing the `presign_loader` which was mostly done to answer perf issues.

So to sum up:
- [ ] Submit to a new branch
- [ ] Test for backward compatibility
- [ ] Performance tests

File: setup.py [hunk 4]
```diff
@@ -49,19 +49,19 @@ def readme():
         "Operating System :: OS Independent",
         "Topic :: Software Development :: Libraries :: Python Modules",
         "Topic :: Internet",
-        "Programming Language :: Python :: 2",
-        "Programming Language :: Python :: 2.7",
+        "Programming Language :: Python :: 3",
```
Comment (amanagr) [inline setup.py]: Please add Python 3.6 support too.

File: tc_aws/storages/s3_storage.py (hunk 2 above)
Comment (kkopachev) [reply tc_aws/storages/s3_storage.py:147]: Not sure. This forced me thinking on what to do if there is an exception while reading from s3. Should code catch all external exceptions, log them and return None? or should it throw read exceptions to callers?

File: tests/test_result_storage.py (hunk 3 above)
Comment (kkopachev) [reply tests/test_result_storage.py:29]: They are covered by next 2 tests, which store and get images in one shot.

**Comment by kkopachev** [comment]
I saw aiobotocore was updated to v1.0 and decided to update it here. They introduced braking change in order to make credentials refreshing asynchronous. Changed is that `create_client` method is now has to be called as context manager.
If we're going to use it that way, then client would be recreated for each outgoing request to s3. Client would also create a pool of aio http connections and kill it right after request. 
I am struggling to find a decent way to make one client and reuse it (singleton) for the life of thumbor. That's what Bucket object was meant for. 
Credentials would live on a shared session object, there is no problem sharing them across requests. The problem is with clients and http pools.
I [opened an issue with aiobotocore](https://github.com/aio-libs/aiobotocore/issues/787) to solicit some guidance on the situation. 
One possible solution would be to call `__aenter__`/`__aexit__` methods manually and store client in a loader/storage instance property, but that looks very magical.

Would love if somebody here can also take a look.

**Comment by heynemann** [comment]
I would comment but have zero context in aiobotocore. :(

**Comment by amanagr** [comment]
Hey, Just wanted you guys to know that I am excited about this PR since we in https://github.com/zulip use this package and it's a potential blocker for us to upgrade thumbor from python 2 to 3. Let me know if I can help in any way.

**Comment by timabbott** [comment]
Any update on this?  It's blocking Zulip testing the upgrade to Thumbor 3.

**Comment by kkopachev** [comment]
Hey guys, I've pushed changes to accommodate latest aiobotocore version. Take a look please.
I know we should be closing the connection pool on thumbor exit, but that does not seem to be an option, so I left it open, hoping python will clean up all remaining resources on exit. Might be not ideal solution. 

@Bladrak I've submitted to new branch. Backward compat should be covered by tests which had rather small changes. Performance should probably be tested, not sure how exactly. I've removed presigning loader since in theory it should not be faster, since all aws requests are already async. 
Also theoretically this branch should be faster in case using instance metadata or ecs task metadata to fetch credentials, since such request (and refresh request too) would now be async.

Please CR.

@heynemann think it's worth adding async initializer/finalizer to base storages to be called on thumbor startup/shutdown?

File: Readme.md [hunk 5]
```diff
@@ -36,7 +36,6 @@ We recommend to use python-virtualevn (virtualenv and virtualenv-wrapper)
 ## Features
```
Comment (Bladrak) [inline Readme.md:36]: We could probably use a Disclaimer section for the branch (ie: beta branch, please test and report any issue you find), as well as a BC section for the removal of presigning loader

File: tests/test_result_storage.py (hunk 3 above)
Comment (Bladrak) [reply tests/test_result_storage.py:29]: It can be interesting to have specific tests (unitary) to ease out bug resolution. However this may be factorized.

**Comment by amanagr** [comment]
I appreciate the progress, but we are blocked on this PR before upgrading to python3 for thumbor. We would appreciate its completion.

**Comment by Bladrak** [comment]
@amanagr if you want to help, a good way might be to test it on your side. Once we have some feedback we may release it as an alpha.

**Comment by amanagr** [comment]
I am trying to test this on a s3 bucket, I received the following error when uploading the image:

```
tc_aws/aws/bucket.py", line 61, in get
    return await self._client.get_object(
AttributeError: 'ClientCreatorContext' object has no attribute 'get_object'
```

Any idea what might be wrong?

**Comment by amanagr** [comment]
I am not sure what you guys are doing here:
```
            _cm_type = type(cm)

            self._client = await _cm_type.__aenter__(cm)
```

why not directly use cm?

**Comment by amanagr** [comment]
I was trying to just upgrade this project from python2 to python3 without the move to aiobotocore. Then I saw https://github.com/nanvel/tornado-botocore/issues/18 and I wonder if the move to aiobotocore is necessary since tornado-botocore is deprecated and doesn't play nice with the latest dependencies in the thumbor ecosystem (thumbor + libthumbor).

**Comment by Bladrak** [comment]
Closing in favor of #147
